src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,556 @@
+#include <sstream>
+#include <utility>
+#include <vector>
+#include <cstdlib>
+
+namespace stan {
+namespace mcmc {
+using Eigen::Dynamic;
+
+/**
+ * An <code>mcmc::chainset</code> object manages the post-warmup draws
+ * across a set of MCMC chains, which all have the same number or samples.
+ *
+ * <p><b>Storage Order</b>: Storage is column/last-index major.
+ */
+class chainset {
+ private:
+  size_t num_samples_;
+  std::vector<std::string> param_names_;
+  std::vector<Eigen::MatrixXd> chains_;
+
+  static size_t thinned_samples(const stan::io::stan_csv& stan_csv) {
+    size_t thinned_samples = stan_csv.metadata.num_samples;
+    if (stan_csv.metadata.thin > 0) {
+      thinned_samples = thinned_samples / stan_csv.metadata.thin;
+    }
+    return thinned_samples;
+  }
+
+  static bool is_valid(const stan::io::stan_csv& stan_csv) {
+    if (stan_csv.header.empty())
+      return false;
+    if (stan_csv.samples.size() == 0)
+      return false;
+    if (stan_csv.samples.rows() != thinned_samples(stan_csv))
+      return false;
+    return true;
+  }
+
+  /**
+   * Process first chain: record header, thinned samples,
+   * add samples to vector chains.
+   */
+  void init_from_stan_csv(const stan::io::stan_csv& stan_csv) {
+    if (!is_valid(stan_csv)) {
+      throw std::invalid_argument("Invalid sample");
+    }
+    if (chains_.size() > 0) {
+      throw std::invalid_argument("Cannot re-initialize chains object");
+    }
+    param_names_ = stan_csv.header;
+    num_samples_ = thinned_samples(stan_csv);
+    chains_.push_back(stan_csv.samples);
+  }
+
+  /**
+   * Process next chain: validate size, shape, column names,
+   * append to vector chains.
+   */
+  void add(const stan::io::stan_csv& stan_csv) {
+    if (!is_valid(stan_csv)) {
+      throw std::invalid_argument("Invalid sample");
+    }
+    if (stan_csv.header.size() != num_params()) {
+      throw std::invalid_argument(
+          "Error add(stan_csv): number of columns in"
+          " sample does not match first chain");
+    }
+    if (thinned_samples(stan_csv) != num_samples_) {
+      throw std::invalid_argument(
+          "Error add(stan_csv): number of sampling draws in"
+          " sample does not match first chain");
+    }
+    for (int i = 0; i < num_params(); i++) {
+      if (param_names_[i] != stan_csv.header[i]) {
+        std::stringstream ss;
+        ss << "Error add(stan_csv): header " << param_names_[i]
+           << " does not match chain's header (" << stan_csv.header[i] << ")";
+        throw std::invalid_argument(ss.str());
+      }
+    }
+    chains_.push_back(stan_csv.samples);
+  }
+
+ public:
+  explicit chainset(const stan::io::stan_csv& stan_csv) {
+    init_from_stan_csv(stan_csv);
+  }
+
+  explicit chainset(const std::vector<stan::io::stan_csv>& stan_csv) {
+    if (stan_csv.empty())
+      return;
+    init_from_stan_csv(stan_csv[0]);
+    for (size_t i = 1; i < stan_csv.size(); ++i) {
+      add(stan_csv[i]);
+    }
+  }
```

Comment:
Can we write a `thinned_samples` function with a signature `thinned_samples(const std::vector<stan::io::stan_csv>&)`? Also add a function like the below so that we can use a member initialization list

```c++
std::vector<Eigen::MatrixXd> extract_samples(const std::vector<stan::io::stan_csv>&)`
```

```suggestion
  explicit chainset(const std::vector<stan::io::stan_csv>& stan_csvs) :
    param_names_(stan_csvs[0].header),
    num_samples_(thinned_samples(stan_csvs)),
    chains_(extract_samples(stan_csvs)) {}
```